cloud-sql/sql-server/sqlalchemy/main.py

```diff
@@ -0,0 +1,170 @@
+            Column("vote_id", Integer, primary_key=True, nullable=False),
+            Column("time_cast", DateTime, nullable=False),
+            Column("candidate", String(6), nullable=False),
+        )
+        metadata.create_all()
+
+
+@app.route("/", methods=["GET"])
+def index():
+    votes = []
+    with db.connect() as conn:
+        # Execute the query and fetch all results
+        recent_votes = conn.execute(
+            "SELECT TOP(5) candidate, time_cast FROM votes "
+            "ORDER BY time_cast DESC"
+        ).fetchall()
+        # Convert the results into a list of dicts representing votes
+        for row in recent_votes:
+            votes.append({"candidate": row[0], "time_cast": row[1]})
+
+        stmt = sqlalchemy.text(
+            "SELECT COUNT(vote_id) FROM votes WHERE candidate=:candidate"
+        )
+        # Count number of votes for tabs
+        tab_result = conn.execute(stmt, candidate="TABS").fetchone()
+        tab_count = tab_result[0]
+        # Count number of votes for spaces
+        space_result = conn.execute(stmt, candidate="SPACES").fetchone()
+        space_count = space_result[0]
+
+    return render_template("index.html", recent_votes=votes,
+                           tab_count=tab_count, space_count=space_count)
+
+
+@app.route("/", methods=["POST"])
+def save_vote():
+    # Get the team and time the vote was cast.
+    team = request.form["team"]
+    time_cast = datetime.datetime.utcnow()
+    # Verify that the team is one of the allowed options
+    if team != "TABS" and team != "SPACES":
+        logger.warning(team)
+        return Response(response="Invalid team specified.", status=400)
+
+    # [START cloud_sql_server_python_sqlalchemy_connection]
+    # Preparing a statement before hand can help protect against injections.
+    stmt = sqlalchemy.text(
+        "INSERT INTO votes (time_cast, candidate)"
+        " VALUES (:time_cast, :candidate)"
+    )
+    try:
+        # Using a with statement ensures that the connection is always released
+        # back into the pool at the end of statement (even if an error occurs)
+        with db.connect() as conn:
+            conn.execute(stmt, time_cast=time_cast, candidate=team)
+    except Exception as e:
+        # If something goes wrong, handle the error in this section. This might
+        # involve retrying or adjusting parameters depending on the situation.
+        # [START_EXCLUDE]
+        logger.exception(e)
+        return Response(
+            status=500,
+            response="Unable to successfully cast vote! Please check the "
+            "application logs for more details.",
+        )
+        # [END_EXCLUDE]
+    # [END cloud_sql_server_python_sqlalchemy_connection]
+
+    return Response(
+        status=200,
+        response="Vote successfully cast for '{}' at time {}!".format(
+                    team, time_cast),
+    )
+
+
+if __name__ == "__main__":
+    app.run(host="127.0.0.1", port=8080, debug=True)
```

Comment:
Should this listen to $PORT?

Reply:
The GAE flex docs only said to listen on 8080, so I'm not sure if $PORT is defined like it is in Cloud Run 
https://cloud.google.com/appengine/docs/flexible/custom-runtimes/build